go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -204,6 +222,63 @@ func (ncc *networkClusterController) init() error {
 	return nil
 }
 
+// updateNetworkStatus allows to report an status for networkClusterController's network via a UDN status condition
+// of type "NetworkAllocationSucceeded", if the network was created by UDN.
+// When at least one node reports an error, condition will be set to false and an event with node-specific error will be
+// generated.
+// Call this function after every node event handling, set errorMsg to empty string to report no error.
+func (ncc *networkClusterController) updateNetworkStatus(nodeName string, errorMsg string) {
+	if ncc.statusReporter == nil {
+		return
+	}
+
+	ncc.nodeErrorsLock.Lock()
+	defer ncc.nodeErrorsLock.Unlock()
+	if ncc.nodeErrors[nodeName] == errorMsg {
+		// error message didn't change for that node, no need to update
+		return
+	}
+	if errorMsg == "" {
+		delete(ncc.nodeErrors, nodeName)
+	} else {
+		ncc.nodeErrors[nodeName] = errorMsg
+	}
+	ncc.reportStatus()
+}
+
+// must be called with ncc.nodeErrorsLock held
+func (ncc *networkClusterController) reportStatus() {
+	if ncc.statusReporter == nil {
+		return
+	}
+	netName := ncc.NetInfo.GetNetworkName()
+	condition := metav1.Condition{
+		Type:               "NetworkAllocationSucceeded",
+		LastTransitionTime: metav1.Now(),
+	}
+	events := make([]*util.EventDetails, 0, len(ncc.nodeErrors))
+	if len(ncc.nodeErrors) == 0 {
+		condition.Status = metav1.ConditionTrue
+		condition.Reason = "NetworkAllocationSucceeded"
+		condition.Message = "Network allocation succeeded for all synced nodes."
+	} else {
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = "InternalError"
+		condition.Message = "Network allocation failed for one or more nodes, check UDN events for more info."
+		for nodeName, errMsg := range ncc.nodeErrors {
+			events = append(events, &util.EventDetails{
+				EventType: util.EventTypeWarning,
+				Reason:    "NetworkAllocationFailed",
+				Note:      fmt.Sprintf("Error occurred for node %s: %s", nodeName, errMsg),
+			})
```

Comment:
Why would we accumulate all the node errors and for each error, send a batch of events with all of the errors?

So for example node A fail, and we sent an event for node A. Then node B fails, and we send an event for node A and B, so now we sent two events for node A.

vs

When node A fail, and we sent an event for node A. Then node B fails, and we send an event for node B only.